src/rgw/rgw_flight.cc

```diff
@@ -0,0 +1,735 @@
+
+#include <iostream>
+#include <fstream>
+#include <mutex>
+#include <map>
+#include <algorithm>
+
+#include "arrow/type.h"
+#include "arrow/buffer.h"
+#include "arrow/util/string_view.h"
+#include "arrow/io/interfaces.h"
+#include "arrow/ipc/reader.h"
+#include "arrow/table.h"
+
+#include "arrow/flight/server.h"
+
+#include "parquet/arrow/reader.h"
+
+#include "common/dout.h"
+#include "rgw_op.h"
+
+#include "rgw_flight.h"
+#include "rgw_flight_frontend.h"
+
+
+#define dout_subsys ceph_subsys_rgw_flight
+
+
+namespace rgw::flight {
+
+  // Ticket and FlightKey
+
+  std::atomic<FlightKey> next_flight_key = null_flight_key;
+
+  flt::Ticket FlightKeyToTicket(const FlightKey& key) {
+    flt::Ticket result;
+    result.ticket = std::to_string(key);
+    return result;
+  }
+
+  arw::Result<FlightKey> TicketToFlightKey(const flt::Ticket& t) {
+    try {
+      return (FlightKey) std::stoul(t.ticket);
+    } catch (std::invalid_argument const& ex) {
+      return arw::Status::Invalid(
+	"could not convert Ticket containing \"%s\" into a Flight Key",
+	t.ticket);
+    } catch (const std::out_of_range& ex) {
+      return arw::Status::Invalid(
+	"could not convert Ticket containing \"%s\" into a Flight Key due to range",
+	t.ticket);
+    }
+  }
+
+  // FlightData
+
+  FlightData::FlightData(const std::string& _uri,
+			 const std::string& _tenant_name,
+			 const std::string& _bucket_name,
+			 const rgw_obj_key& _object_key,
+			 uint64_t _num_records,
+			 uint64_t _obj_size,
+			 std::shared_ptr<arw::Schema>& _schema,
+			 std::shared_ptr<const arw::KeyValueMetadata>& _kv_metadata,
+			 rgw_user _user_id) :
+    key(++next_flight_key),
+    /* expires(coarse_real_clock::now() + lifespan), */
+    uri(_uri),
+    tenant_name(_tenant_name),
+    bucket_name(_bucket_name),
+    object_key(_object_key),
+    num_records(_num_records),
+    obj_size(_obj_size),
+    schema(_schema),
+    kv_metadata(_kv_metadata),
+    user_id(_user_id)
+  { }
+
+  /**** FlightStore ****/
+
+  FlightStore::FlightStore(boost::intrusive_ptr<ceph::common::CephContext>& _cct) :
+    cct(_cct),
+    dp(cct.get(), dout_subsys, "rgw arrow_flight: ")
+  { }
+
+  FlightStore::~FlightStore() { }
+
+  /**** MemoryFlightStore ****/
+
+  MemoryFlightStore::MemoryFlightStore(boost::intrusive_ptr<ceph::common::CephContext>& _cct) :
+    FlightStore(_cct)
+  { }
+
+  MemoryFlightStore::~MemoryFlightStore() { }
+
+  FlightKey MemoryFlightStore::add_flight(FlightData&& flight) {
+    std::pair<decltype(map)::iterator,bool> result;
+    {
+      const std::lock_guard<decltype(mtx)> lock(mtx);
+      result = map.insert( {flight.key, std::move(flight)} );
+    }
+    ceph_assertf(result.second,
+		 "unable to add FlightData to MemoryFlightStore"); // temporary until error handling
+
+    return result.first->second.key;
+  }
+
+  arw::Result<FlightData> MemoryFlightStore::get_flight(const FlightKey& key) const {
+    const std::lock_guard<decltype(mtx)> lock(mtx);
```

Comment:
since c++17, you can just write these as `const std::lock_guard lock(mtx);`. see https://en.cppreference.com/w/cpp/language/class_template_argument_deduction